### DISCUSSION TITLE
Enhanced RTMP & FLV test/sample media, clients, and servers for validation and interop

### DISCUSSION THREAD

**Discussion by zenomt** [veovera/enhanced-rtmp#19] [General]
I've been adding Enhanced RTMP support to my implementations.

Do folks have test & sample media in the new enhanced FLV format people can use for validation? Are there compatible clients (RTMP viewers & publishers) and servers with which to validate?

**Comment by zenomt** [comment]
[My RTMP/RTMFP server "tcserver"](https://github.com/zenomt/rtmfp-cpp/blob/main/test/tcserver.md) _should_ be compatible with Enhanced RTMP (& RTMFP). Likewise [my "tcpublish" tool](https://github.com/zenomt/rtmfp-cpp/blob/main/test/tcpublish.cpp) _should_ correctly publish an enhanced FLV to an enhanced-compatible server (though tcpublish only speaks RTMFP).

I'd like to be able to validate this though. :)

**Comment by zenomt** [comment]
And I'm sure the community would like to be able to test interoperation between different tools, clients, servers, services, etc.

**Comment by rogerhu** [comment]
I'm also close to getting changes landed in this tool as well: https://github.com/noirotm/flvmeta/pull/20

**Comment by zenomt** [comment]
with the release of a new, updated Enhanced RTMP spec now with new audio codecs (among other enhancements), i'm renewing my request for either:
* sample FLV media in this repo that demonstrates the new features and codecs
* links in this repo to sample media elsewhere that demonstrates the new features and codecs

sample media is essential for ensuring interoperability among implementations.

**Reply by veovera** [reply]
Did you ever get your hands on some sample media? I am asking around as well...

**Reply by zenomt** [reply]
i have seen no sample media for the new enhanced audio.

i was given private access to HEVC, AV1, and VP9 media a year ago, but that doesn't do anyone else any good. :)

**Reply by zenomt** [reply]
we really need a public library of sample media.

**Reply by jhnbwrs** [reply]
I have some twitch sample content I have created here: https://d50yg09cghihd.cloudfront.net/sample_eflv_content/eflv_sample_assets.txt

**Reply by veovera** [reply]
@jhnbwrs This is awesome and thank you for sharing!

**Reply by ThibaultBee** [reply]
@jhnbwrs I had a quick look at your test file. The `onMetaData` tag is wrapped within `@setDataFrame`, which, if I'm not mistaken, is not valid FLV syntax (but a RTMP syntax).

**Reply by zenomt** [reply]
@ThibaultBee i believe you're correct. `onMetaData` should be plain in an FLV. something publishing an FLV to an RTMP server _should_ wrap `onMetaData` (and anything else it knows that should be a data keyframe) in an `@setDataFrame` -- for example: https://github.com/zenomt/rtmfp-cpp/blob/2dfe10880f2438db1ec5582697463bcadc23866a/test/tcpublish.cpp#L417-L418 :)

**Reply by zenomt** [reply]
note however that it's perfectly _valid_ to have an `@setDataFrame` scriptdata tag in an FLV, because that's just another method name. but it likely won't have the desired effect unless you're sending it to an RTMP server that implements "data keyframes" like Adobe Media Server does.

**Comment by koolscooby** [comment]
A few other resources for folks :

* ertmp-toolbox: Simple rust utility that can read and push eFLV files https://github.com/kc5nra/ertmp-toolbox
* FFmpeg: Branch that has Enhanced RTMP Multitrack FLV support https://github.com/BtbN/FFmpeg/tree/enhanced-flv
* Wireshark: v4.4+ support dissecting/parsing Enhanced RTMP Multitrack connections
* OBS Studio: v30.2+ support Twitch Enhanced Broadcasting powered by Enhanced RTMP Multitrack Feature. A CLI option --enable-multitrack-video-dev lets you experiment with custom OBS configurations for testing streaming to services other than Twitch and provides an optional checkbox to record an eFLV file.

**Reply by veovera** [reply]
I started a file that keeps track for all the resources https://github.com/veovera/enhanced-rtmp/blob/main/docs/news/resources.md. Feel free to suggest to add entries!

@zenomt I want to point to your project from the file. What can I say for the link description?

**Reply by zenomt** [reply]
since i'm not currently supporting multitrack, you could say this:
> **rtmfp-cpp**: C++ RTMFP and RTMP implementations (with example live media server) supporting new Enhanced RTMP video and audio codecs. [GitHub Link](https://github.com/zenomt/rtmfp-cpp)

and as long as you're collecting useful "implementation resources" from third parties, that page seems like a good place to link to my Errata and Addenda doc (to help people with the regular parts of RTMP):
> [**RTMP Errata and Addenda**](https://zenomt.github.io/rtmp-errata-addenda/): Corrections and clarifications to the legacy RTMP specification. [HTML](https://zenomt.github.io/rtmp-errata-addenda/rtmp-errata-addenda.html) | [Plain Text](https://zenomt.github.io/rtmp-errata-addenda/rtmp-errata-addenda.txt)

**Reply by zenomt** [reply]
@veovera it looks like this got dropped. were you still interested in linking to my projects from the "resources" page?

**Reply by zenomt** [reply]
and i still haven't implemented high resolution timestamps. but i have a staycation coming up next week and that's on my list to spend some time on. :)
